packages/template-retail-react-app/app/hooks/use-update-shopper-context.js

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import {useEffect} from 'react'
+import {useLocation} from 'react-router-dom'
+import {
+    useUsid,
+    useShopperContext,
+    useShopperContextsMutation
+} from '@salesforce/commerce-sdk-react'
+import {isServer, isHydrated} from '@salesforce/retail-react-app/app/utils/utils'
+import {useQueryClient} from '@tanstack/react-query'
+
+// Hooks
+import useMultiSite from '@salesforce/retail-react-app/app/hooks/use-multi-site'
+
+// Constants
+import {useShopperContextSearchParams} from '@salesforce/retail-react-app/app/hooks/use-shopper-context-search-params'
+
+/*
+ * This hook will set the shopper context when search params pertinant
+ * to shopper context are present.
+ */
+export const useUpdateShopperContext = () => {
+    const {site} = useMultiSite()
+    const {usid} = useUsid()
+    const {search} = useLocation()
+    const queryClient = useQueryClient()
+    const createShopperContext = useShopperContextsMutation('createShopperContext')
+    const updateShopperContext = useShopperContextsMutation('updateShopperContext')
+    const {data: shopperContext} = useShopperContext(
+        {parameters: {usid, siteId: site.id}},
+        {enabled: !isServer}
+    )
+
+    const refetchDataOnClient = () => {
+        queryClient.invalidateQueries()
+    }
+
+    const handleShopperContextUpdate = async (updateShopperContextObj) => {
+        if (Object.keys(updateShopperContextObj).length === 0) {
+            return
+        }
+
+        const payload = {
+            parameters: {usid, siteId: site.id},
+            body: updateShopperContextObj
+        }
+        if (!shopperContext) {
+            await createShopperContext.mutateAsync(payload)
+        } else {
+            await updateShopperContext.mutateAsync(payload)
+        }
+
+        // Refresh data
+        refetchDataOnClient()
+    }
+
+    // Handle updating the shopper context based on URL search params
+    const updateShopperContextObj = useShopperContextSearchParams()
+    useEffect(() => {
+        handleShopperContextUpdate(updateShopperContextObj)
+    }, [search])
```

Comment:
🤔 You knew that `search` here should be the dependency because you're aware of the implementation of `updateShopperContextObj`. But it's not immediately clear to other developers.

Try this instead. The `updateShopperContextObj` value should change according to `search`.
```suggestion
    }, [updateShopperContextObj])
```

Reply:
@vmarta so I originally tried this but it seems that for some reason it'll start going in an endless loop. I think it has to do with `updateShopperContextObj` being an object?

I will try to test if `useMemo` or maybe `JSON.stringify` the object resolves the issue, but open to suggestions!